### ISSUE TITLE
[feature] Python library for omnictl API

### ISSUE THREAD

**Issue by ivancavanaugh** [siderolabs/omni#735]
### Problem Description

Current integration on the gRPC endpoints in omni is only provided in a go-lang library. With use of other systems which are not based on go-lang it is very difficult to integrate to Omni to retrieve information on clusters, machines, etc.

If the omnictl could be also provided in other languages which are also mainstream, like python, java, .net, etc, it would help a lot.

For us it is required to retrieve information from Omni to build metadata around the machine and provide customer application profiles in the form of helm charts specific for that customer. Without information from Omni of which machines and clusters are registered this metadata could not be composed.

### Solution

Convert the Omnictl code from go-lang to python and provide this as a library via pip

**Comment by utkuozdemir** [comment]
Hi,

We discussed the idea of Python client library internally, and we lean more towards an `omnictl` CLI wrapper for that, as our Go client contains quite a bit of non-trivial logic to marshal/unmarshal resource protobufs, sanitize, handle and format input/outputs etc.

Can you elaborate a bit on your use cases? Especially on:

> For us it is required to retrieve information from Omni to build metadata around the machine and provide customer application profiles in the form of helm charts specific for that customer. Without information from Omni of which machines and clusters are registered this metadata could not be composed.

An example use case would be great.

Additionally: AFAIK you are currently using/wrapping the CLI already. Can you clarify the following:
- What are the your pain points when wrapping the CLI? Is it about some missing output formats / consistency? Is it about the exposed features? Or the need to keep it up-to-date?
- About authentication - are you using service account auth for the automations you mentioned?
- Most importantly: are you using cluster templates? Do you want to be able to use it from Python? Which subcommands of it would you need to address your requirement.

**Comment by ivancavanaugh** [comment]
Thank you for your answer and clarification. I do understand that there are difficulties to port the code to python.

A use casse that we have in the metadata system now is as followed:
1. The machines are retrieved from Omni using the Omnictl by executing this in a subprocess in python and capturing the yaml output
2. The clustermachines are retrieved in the same way
3. Using talosctl the device hardware details are retrieved
4. The make and model of the device is determined and the bootstrap files are constructed as a cluster template and sent to the device unsing omnictl again in a subprocess
5. and a lot more of these kind of operations

Having the dependency on the binary executable is in our opinion not an optimal solution. Since the format of the response yaml is not validated the response data needs to be interpreted by the code and that is where things can go wrong. Null references is the biggest part in this. To have a lot of checks in the code does not help in readability and understanding of the requirements and the flow data between Omni and MCP.

Regarding authentication we only use the service account in the backend for performing these kind of operations.

**Comment by ivancavanaugh** [comment]
What also can be done is to use a ctypes in python, but it is required for us to perform indepth investigations on which method to call in the go code. The python library may also be based on this technique, as long as we can use pip to update/upgrade dependencies following your release schedule.

```
from ctypes import *
    def call_go_function():

        lib = cdll.LoadLibrary("./awesome.so")
        lib.Add.argtypes = [c_longlong, c_longlong]
        print( lib.Add(12,99)) 

    call_go_function()   
```

**Comment by smira** [comment]
> What also can be done is to use a ctypes in python, but it is required for us to perform indepth investigations on which method to call in the go code. The python library may also be based on this technique, as long as we can use pip to update/upgrade dependencies following your release schedule.

I don't think you can do this with any Go code due to the way Go runtime works.
